src/plugin/systems/multiple_rapier_contexts.rs

```diff
@@ -0,0 +1,221 @@
+//! systems to support multiple worlds, and changes between them.
+
+use crate::dynamics::{
+    RapierImpulseJointHandle, RapierMultibodyJointHandle, RapierRigidBodyHandle,
+};
+use crate::geometry::RapierColliderHandle;
+use crate::plugin::{RapierContext, RapierContextEntityLink};
+use bevy::prelude::*;
+
+/// If an entity is turned into the child of something with a physics world, the child should become a part of that physics world
+///
+/// If this fails to happen, weirdness will ensue.
+pub fn on_add_entity_with_parent(
+    q_add_entity_without_parent: Query<
+        (Entity, &Parent),
+        (With<RapierContextEntityLink>, Changed<Parent>),
+    >,
+    q_parent: Query<&Parent>,
+    q_physics_world: Query<&RapierContextEntityLink>,
+    mut commands: Commands,
+) {
+    for (ent, parent) in &q_add_entity_without_parent {
+        let mut parent = Some(parent.get());
+        while let Some(parent_entity) = parent {
+            if let Ok(pw) = q_physics_world.get(parent_entity) {
+                commands.entity(ent).insert(*pw);
+                remove_old_physics(ent, &mut commands);
+                break;
+            }
+
+            parent = q_parent.get(parent_entity).ok().map(|x| x.get());
+        }
+    }
+}
+
+/// Flags the entity to have its old physics removed
+fn remove_old_physics(entity: Entity, commands: &mut Commands) {
+    commands
+        .entity(entity)
+        .remove::<RapierColliderHandle>()
+        .remove::<RapierRigidBodyHandle>()
+        .remove::<RapierMultibodyJointHandle>()
+        .remove::<RapierImpulseJointHandle>();
+}
+
+/// Flags the entity to have its physics updated to reflect new world
+///
+/// Also recursively bubbles down world changes to children & flags them to apply any needed physics changes
+pub fn on_change_world(
+    q_changed_worlds: Query<
+        (Entity, Ref<RapierContextEntityLink>),
+        Changed<RapierContextEntityLink>,
+    >,
+    q_children: Query<&Children>,
+    q_physics_world: Query<&RapierContextEntityLink>,
+    q_context: Query<&RapierContext>,
+    mut commands: Commands,
+) {
+    for (entity, new_physics_world) in &q_changed_worlds {
+        let context = q_context.get(new_physics_world.0);
+        if new_physics_world.is_added() {
```

Comment:
My initial mental model was that this function is responsible of reacting to rapier context change ; addition is handled within `init_rigid_bodies` and similar functions.

You're correct though, the initialization is lacking the children propagation, I'll see what I can do.

I think keeping a similar logic is easier to reason about 🤔, rather than making an big system responsible for both initializing and update to link changes.